### PR TITLE
Helpers to change format of ECDSA signature

### PR DESCRIPTION
**Pull request by viktorTarasov**
Format of ECDSA signature, adopted on pkcs15 and pkcs11 levels, is raw concatenation of integers R and S.

 * Some cards supporting EC signature returns computed signature encoded as ASN1 sequence of two integers, and it needs to be transformed into the *raw* format.
 * OpenSSL expects signature encoded as sequence.

These helpers are to be used by tools and card drivers.

### CODE REVIEW

File: src/libopensc/asn1.c [hunk 1]
```diff
@@ -1819,3 +1819,80 @@ sc_encode_oid (struct sc_context *ctx, struct sc_object_id *id,
...
+}
+
+
+int
+sc_asn1_sig_value_sequence_to_rs(struct sc_context *ctx, unsigned char *in, size_t inlen,
+                unsigned char **buf, size_t *buflen)
+{
+        struct sc_asn1_entry asn1_sig_value[C_ASN1_SIG_VALUE_SIZE];
+        struct sc_asn1_entry asn1_sig_value_coefficients[C_ASN1_SIG_VALUE_COEFFICIENTS_SIZE];
+	unsigned char *r, *s;
```
Comment (frankmorgner) [inline src/libopensc/asn1.c]: `asn1.c` uses tabs instead of spaces

File: src/libopensc/asn1.c [hunk 2]
```diff
@@ -1819,3 +1819,80 @@ sc_encode_oid (struct sc_context *ctx, struct sc_object_id *id,
...
+
+        rv = sc_asn1_encode(ctx, asn1_sig_value, buf, buflen);
+        LOG_TEST_RET(ctx, rv, "ASN.1 encoding ECDSA-SIg-Value failed");
+
+        LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+}
+
+
+int
+sc_asn1_sig_value_sequence_to_rs(struct sc_context *ctx, unsigned char *in, size_t inlen,
```
Comment (frankmorgner) [inline src/libopensc/asn1.c]: `sc_asn1_sig_value_sequence_to_rs` and `sc_asn1_sig_value_rs_to_sequence` are application specific and look like they belong to the pkcs#15 layer. If moved, you would not need to add them to `libopensc.exports`.

File: src/pkcs15init/pkcs15-lib.c [hunk 3]
```diff
@@ -2039,12 +2041,18 @@ check_key_compatibility(struct sc_pkcs15_card *p15card, struct sc_pkcs15_prkey *
 			}
 		}
 		else if (key->algorithm == SC_ALGORITHM_EC)   {
+			if (!sc_valid_oid(&key->u.ec.params.id))
+				if (sc_pkcs15_fix_ec_parameters(ctx, &key->u.ec.params))
+					LOG_FUNC_RETURN(ctx, SC_ERROR_OBJECT_NOT_VALID);
+			if (sc_valid_oid(&info->u._ec.curve_oid))
```
Comment (frankmorgner) [inline src/pkcs15init/pkcs15-lib.c]: Is it possible to use custom domain parameters, which don't have an OID. I know a card that can be configured that way, however, in practice it is not used.

Comment (frankmorgner) [inline src/pkcs15init/pkcs15-lib.c]: What I wanted to say is that this check disallows those custom curves.

File: src/libopensc/asn1.c (hunk 2 above)
Comment (viktorTarasov) [inline src/libopensc/asn1.c]: It will be used by card driver.

File: src/pkcs15init/pkcs15-lib.c (hunk 3 above)
Comment (viktorTarasov) [inline src/pkcs15init/pkcs15-lib.c]: I do not see where it disallows a custom curves.
This patch will be active only if card driver declares support of particular curves.

**Comment by philipWendland** [comment]
I tested `sc_asn1_sig_value_sequence_to_rs` - very helpful.

I guess sc_asn1_sig_value_sequence_to_rs() would be used by card drivers in compute_signature(). The current implementation would require to copy the result back to the "out" buffer of compute_signature() and free the buffer allocated by sc_asn1_sig_value_sequence_to_rs. 

From a performance point of view - could `sc_asn1_sig_value_sequence_to_rs` allow "in-place" modification of the data? E.g. by using memmove and requiring the caller to pass a pre-allocated "*buf" buffer as an out parameter.
@dengert This might be interesting for you as well :-)

**Comment by dengert** [comment]
I would skip the in-place code.  It is nothing compared to the time to send the request to the card,
have a card sign it and get the results back.

If it gets committed, I would look at it for the PIV driver.


On 2/18/2015 2:20 PM, Philip Wendland wrote:
> I tested |sc_asn1_sig_value_sequence_to_rs| - very helpful.
>
> I guess sc_asn1_sig_value_sequence_to_rs() would be used by card drivers in compute_signature(). The current implementation would require to copy the result back to the "out" buffer of
> compute_signature() and free the buffer allocated by sc_asn1_sig_value_sequence_to_rs.
>
>  From a performance point of view - could |sc_asn1_sig_value_sequence_to_rs| allow "in-place" modification of the data? E.g. by using memmove and requiring the caller to pass a pre-allocated "*buf"
> buffer as an out parameter.
> @dengert <https://github.com/dengert> This might be interesting for you as well :-)
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/pull/381#issuecomment-74940039>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

File: src/libopensc/asn1.c (hunk 1 above)
Comment (viktorTarasov) [inline src/libopensc/asn1.c]: fixed

File: src/pkcs15init/pkcs15-lib.c (hunk 3 above)
Comment (viktorTarasov) [inline src/pkcs15init/pkcs15-lib.c]: @frankmorgner 
> Is it possible to use custom domain parameters, 

I replaced the curve OID by *EC algorithm info* data to hold the custom domain parameters,
but, afais, in PKCS#11 the use of namedCurve is recommended over the choice ecParameters,
and I suppose that custom domain parameters do not well actually tested/supported in OpenSC.

**Comment by viktorTarasov** [comment]
applied in 80c496671f5bb513a1298b462040ed12d549fb2b
used in b08671fab5b500ed18e7fe189d8e6aadbaf22f4b

**Comment by philipWendland** [comment]
Sorry that I have to get back to this after this has been applied, but I think I found a (rare) issue with this:

```
(performing ECDSA signature operation, data is random data to sign)
Outgoing APDU data [  260 bytes] =====================================
10 2A 9E 9A FF 36 66 61 34 79 64 4E 61 32 35 38 .*...6fa4ydNa258
61 42 6D 75 55 71 44 69 55 6B 59 38 46 65 49 38 aBmuUqDiUkY8FeI8
4F 50 68 69 64 71 4C 59 58 6E 6B 74 4D 52 36 51 OPhidqLYXnktMR6Q
6E 53 56 50 41 42 75 57 79 4C 6E 4B 51 73 75 64 nSVPABuWyLnKQsud
5A 71 5A 49 48 6B 37 44 4D 61 73 33 42 6E 66 5A ZqZIHk7DMas3BnfZ
42 62 6D 55 30 37 78 37 5A 41 32 4E 51 76 6E 45 BbmU07x7ZA2NQvnE
73 57 53 76 33 34 56 54 6D 71 71 42 47 59 68 56 sWSv34VTmqqBGYhV
6F 37 7A 59 34 42 36 71 56 7A 6D 31 69 41 68 6C o7zY4B6qVzm1iAhl
73 35 5A 32 4A 38 44 30 54 66 33 4C 51 79 51 5A s5Z2J8D0Tf3LQyQZ
46 43 7A 35 44 65 52 68 47 6A 6C 4C 52 77 39 6D FCz5DeRhGjlLRw9m
71 69 6E 49 52 6F 48 54 69 5A 67 63 4A 59 61 46 qinIRoHTiZgcJYaF
52 31 33 70 74 7A 4E 51 6E 38 73 51 57 30 56 71 R13ptzNQn8sQW0Vq
56 76 4C 6A 59 47 6B 74 5A 51 4F 4B 43 53 49 6F VvLjYGktZQOKCSIo
53 65 66 74 65 47 61 55 50 70 62 4B 44 4B 33 68 SefteGaUPpbKDK3h
72 41 55 33 66 41 69 41 46 44 69 6B 32 62 61 4C rAU3fAiAFDik2baL
39 6C 65 49 51 52 4D 4C 6D 78 7A 4C 6E 64 4E 72 9leIQRMLmxzLndNr
68 30 6B 62                                     h0kb
======================================================================
0x7f3a48fb5700 20:06:31.504 [opensc-pkcs11] reader-pcsc.c:182:pcsc_internal_transmit: called
0x7f3a48fb5700 20:06:31.564 [opensc-pkcs11] apdu.c:185:sc_apdu_log: 
Incoming APDU data [    2 bytes] =====================================
90 00 ..
======================================================================
0x7f3a48fb5700 20:06:31.564 [opensc-pkcs11] apdu.c:397:sc_single_transmit: returning with: 0 (Success)
0x7f3a48fb5700 20:06:31.564 [opensc-pkcs11] apdu.c:550:sc_transmit: returning with: 0 (Success)
0x7f3a48fb5700 20:06:31.564 [opensc-pkcs11] apdu.c:528:sc_transmit: called
0x7f3a48fb5700 20:06:31.564 [opensc-pkcs11] apdu.c:382:sc_single_transmit: called
0x7f3a48fb5700 20:06:31.564 [opensc-pkcs11] apdu.c:387:sc_single_transmit: CLA:0, INS:2A, P1:9E, P2:9A, data(6) 0x7fffd79eb21f
0x7f3a48fb5700 20:06:31.564 [opensc-pkcs11] reader-pcsc.c:249:pcsc_transmit: reader '*'
0x7f3a48fb5700 20:06:31.564 [opensc-pkcs11] apdu.c:185:sc_apdu_log: 
Outgoing APDU data [   12 bytes] =====================================
00 2A 9E 9A 06 56 4C 6B 6F 7A 0A 00 .*...VLkoz..
======================================================================
0x7f3a48fb5700 20:06:31.564 [opensc-pkcs11] reader-pcsc.c:182:pcsc_internal_transmit: called
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] apdu.c:185:sc_apdu_log: 
Incoming APDU data [   72 bytes] =====================================
30 44 02 20 59 0A 21 65 63 BF 8A 01 61 BA 00 4F 0D. Y.!ec...a..O
DB C6 15 AC 9C 63 A7 A1 56 22 CD EF 70 66 2F FD .....c..V"..pf/.
9B 98 61 38 02 20 00 91 2E 23 12 3A AE 13 0F 1F ..a8. ...#.:....
2F 98 20 C1 CC 67 DB 74 6B F5 5F 3A 6F CE C7 2D /. ..g.tk._:o..-
17 C5 A2 3B 6D DF 90 00                         ...;m...
======================================================================
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] apdu.c:397:sc_single_transmit: returning with: 0 (Success)
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] apdu.c:550:sc_transmit: returning with: 0 (Success)
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] card.c:385:sc_unlock: called
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] iso7816.c:916:iso7816_compute_signature: returning with: 70
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] asn1.c:1876:sc_asn1_sig_value_sequence_to_rs: called
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] asn1.c:1441:asn1_decode: called, left=70, depth 0
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] asn1.c:1465:asn1_decode: Looking for 'ECDSA-Sig-Value', tag 0x1000010
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] asn1.c:1257:asn1_decode_entry: decoding 'ECDSA-Sig-Value'
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] asn1.c:1441:asn1_decode:  called, left=68, depth 1
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] asn1.c:1465:asn1_decode: Looking for 'r', tag 0x2
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] asn1.c:1257:asn1_decode_entry:  decoding 'r'
0x7f3a48fb5700 20:06:31.709 [opensc-pkcs11] asn1.c:1465:asn1_decode: Looking for 's', tag 0x2
0x7f3a48fb5700 20:06:31.710 [opensc-pkcs11] asn1.c:1257:asn1_decode_entry:  decoding 's'
0x7f3a48fb5700 20:06:31.710 [opensc-pkcs11] asn1.c:1514:asn1_decode: returning with: 0 (Success)
0x7f3a48fb5700 20:06:31.710 [opensc-pkcs11] asn1.c:1514:asn1_decode: returning with: 0 (Success)
0x7f3a48fb5700 20:06:31.710 [opensc-pkcs11] asn1.c:1887:sc_asn1_sig_value_sequence_to_rs: r(32): 590A216563BF8A0161BA004FDBC615AC 9C63A7A15622CDEF70662FFD9B986138
0x7f3a48fb5700 20:06:31.710 [opensc-pkcs11] asn1.c:1888:sc_asn1_sig_value_sequence_to_rs: s(31): 912E23123AAE130F1F2F9820C1CC67DB 746BF55F3A6FCEC72D17C5A23B6DDF
```

Have a look at the second (S) component returned by the card:
```
02 20 
00 91 2E 23 12 3A AE 13 0F 1F 2F 98 20 C1 CC 67 DB 74 6B F5 5F 3A 6F CE C7 2D 17 C5 A2 3B 6D DF
```
There is a leading `00` that is not *added* to force the number to be positive in the ASN1 sequence. But is removed by the asn1 helper function, which is correct according to the ASN1 logic. Thus, the written signature is one byte short on the S component. The verification will fail because of this. 

Statsistically, this error will occur in very few cases - when the first byte of one or more components (R,S) is `00` and was *not added* to force the integer to be positive. Still, it is an issue...

**Comment by viktorTarasov** [comment]
Thank you,
fixed in 8cf99a93726e735e04ed2563f14e315a60a4a859.

**Comment by dengert** [comment]
The patch may not work if both r and s have leading zero bytes, because nLen is
not know to the routine, in order to add the leading zero bytes back in. 

PKCS#11 2.20 12.3.1 EC Signatures, discribes the problem we are having:
"For signatures created by a token, the resulting signature is always of length 2nLen."
and:
"Note: For applications, it is recommended to encode the signature as an octet string of
length two times nLen if possible. This ensures that the application works with PKCS#11
modules which have been implemented based on an older version of this document.
Older versions required all signatures to have length two times nLen. It may be
impossible to encode the signature with the maximum length of two times nLen if the
application just gets the integer values of r and s (i.e. without leading zeros), but does not
know the base point order n, because r and s can have any value between zero and the
base point order n."

**Pull request closed without merging**